chain/chain/src/chain.rs

```diff
@@ -4367,7 +4367,7 @@ impl<'a> ChainUpdate<'a> {
     /// to process the block an the block is valid.
     //  Note that this function does NOT introduce any changes to chain state.
     fn preprocess_block(
-        &mut self,
+        &self,
```

Comment:
cc @mzhangmzz I think you'll like this! 

The fact that we only need `&` means that we can move the whole thing to chain, where it seems it belongs!